lib/spack/spack/platforms/cray.py

```diff
@@ -118,17 +126,28 @@ def _default_target_from_env(self):
         # env -i /bin/bash -lc echo $CRAY_CPU_TARGET 2> /dev/null
         if getattr(self, 'default', None) is None:
             env = which('env')
-            output = env("-i", "/bin/bash", "-lc", "echo $CRAY_CPU_TARGET",
-                         output=str, error=os.devnull)
-            self.default = output.strip()
-            tty.debug("Found default module:%s" % self.default)
-        return self.default
+            output = env("-i", "/bin/bash", "--norc", "--noprofile", "-lc",
+                         "echo $CRAY_CPU_TARGET", output=str, error=os.devnull)
+            default_from_module = output.strip()
+            if default_from_module:
+                tty.debug("Found default module:%s" % default_from_module)
+                return default_from_module
+            else:
+                front_end = cpu.host().name
+                if front_end in list(
+                        map(lambda x: _target_name_from_craype_target_name(x),
+                            self._avail_targets())
+                ):
+                    tty.debug("default to front-end architecture")
+                    return cpu.host().name
+                else:
+                    return platform.machine()
 
     def _avail_targets(self):
         '''Return a list of available CrayPE CPU targets.'''
         if getattr(self, '_craype_targets', None) is None:
             output = module('avail', '-t', 'craype-')
             craype_modules = _get_modules_in_modulecmd_output(output)
             self._craype_targets = targets = []
-            _fill_craype_targets_from_modules(targets, craype_modules)
+            _fill_craype_target_names_from_modules(targets, craype_modules)
```

Comment:
Above we assume that `module -t av craype-` will return something sensible, but it might not be the case after a `module purge`. Is it worth as a backup option to look at the name of files in locations like:
```
/opt/cray/pe/craype/X.Y.Z/modulefiles
```
or would it be a problem for any reason (too much stress on the filesystem / location changes depending on Cray machine, etc. )?

Also, would it be worth in the long term to cache somewhere on file the result of this detection and refresh the cache every so often (or on demand)?

Reply:
I would lean away from a file cache because we could be operating out of a directory cross-mounted on two machines.